grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+                ExitStatus::Signal(sig) => match sig {
+                    SignalType::SegFault => {
+                        run_updates.add_update(
+                            Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                                .notes("Your submission was killed by SIGSEGV!"),
+                        );
+                        results_sect.add_content(log);
+                        sect.add_content(run_updates);
+                        sect.add_content(Content::SubSection(results_sect));
+                    }
+                    SignalType::Abort => {
+                        run_updates.add_update(
+                            Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                                .notes("Your submission was killed by SIGABRT!"),
+                        );
+                        results_sect.add_content(log);
+                        sect.add_content(run_updates);
+                        sect.add_content(Content::SubSection(results_sect));
+                    }
+                },
+                _ => panic!("Expected either Timeout or Signal if command was not completed"),
+            }
+            return Ok(StageResult::new(
+                StageStatus::UnrecoverableFailure,
+                Some(Output::new().section(sect)),
+            ));
+        }
+
+        match res.status {
+            ExitStatus::Ok => {
+                run_updates.add_update(Update::new_pass("running valgrind"));
+                results_sect.add_content(log);
+                sect.add_content(run_updates);
+                sect.add_content(Content::SubSection(results_sect));
+
+                Ok(StageResult::new(
+                    StageStatus::Continue {
+                        points_lost: PointQuantity::zero(),
+                    },
+                    Some(Output::new().section(sect)),
+                ))
+            }
+            ExitStatus::Failure(rc) => {
+                if rc < ERROR_EXITCODE {
+                    return Err(anyhow!(
+                        "Error: expected exit code to be at least {}, was {}",
+                        ERROR_EXITCODE,
+                        rc
+                    ));
+                }
+
+                run_updates.add_update(
+                    Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                        .notes("Valgrind Errors Detected"),
+                );
+                results_sect.add_content(log);
+                sect.add_content(run_updates);
+                sect.add_content(Content::SubSection(results_sect));
+
+                return Ok(StageResult::new(
+                    StageStatus::UnrecoverableFailure,
+                    Some(Output::new().section(sect)),
+                ));
+            }
+            _ => panic!("Expected either Ok or Failure ExitStatus if the command completed"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I would have also liked to see a test which compiles a short C program which leaks memory and see if the valgrind stage could detect it correctly. Also interested in how valgrind stage handles programs that abort/segfault as I mentioned above. No need to do that here if you don't want to. We can handle this in integration tests once I get those up and running.

Reply:
Yeah I wasn't sure if those tests belonged in this file, but I can certainly do that

Reply:
If not here, you can wait a bit and we can add as part of integration testing later.